src/langmem/short_term/summarization.py

```diff
@@ -0,0 +1,208 @@
+from dataclasses import dataclass
+from typing import Callable, cast
+
+from langchain_core.language_models.chat_models import BaseChatModel
+from langchain_core.messages import AIMessage, BaseMessage, HumanMessage, SystemMessage
+from langchain_core.prompts.chat import ChatPromptTemplate, ChatPromptValue
+
+TokenCounter = Callable[[list[BaseMessage]], int]
+
+
+DEFAULT_INITIAL_SUMMARY_PROMPT = ChatPromptTemplate.from_messages(
+    [
+        ("placeholder", "{messages}"),
+        ("user", "Create a summary of the conversation above:"),
+    ]
+)
+
+
+DEFAULT_EXISTING_SUMMARY_PROMPT = ChatPromptTemplate.from_messages(
+    [
+        ("placeholder", "{messages}"),
+        (
+            "user",
+            "This is summary of the conversation to date: {existing_summary}\n\n"
+            "Extend the summary by taking into account the new messages above:",
+        ),
+    ]
+)
+
+DEFAULT_FINAL_SUMMARY_PROMPT = ChatPromptTemplate.from_messages(
+    [
+        # if exists
+        ("placeholder", "{system_message}"),
+        ("system", "Summary of conversation earlier: {summary}"),
+        ("placeholder", "{messages}"),
+    ]
+)
+
+
+@dataclass
+class SummaryInfo:
+    # the summary of the conversation so far
+    summary: str
+    # the IDs of the messages that have been most recently summarized
+    summarized_message_ids: list[str]
+    # keep track of the total number of messages that have been summarized thus far
+    total_summarized_messages: int = 0
+
+
+@dataclass
+class SummarizationResult:
+    # the messages that will be returned to the user
+    messages: list[BaseMessage]
+    # SummaryInfo (empty if messages were not summarized)
```

Comment:
`SummaryInfo` is the name of a variable. Generally doc-strings should not reference variable names, but conceptual names
---

(as horrendous as it is) the convention in all of langchain is for the doc-string of the attribute to be below the attribute rather than above